### ISSUE TITLE
Bug: [unified-signatures] False negatives when signatures are identical

### ISSUE THREAD

**Issue by StyleShit** [typescript-eslint/typescript-eslint#12670]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&esQuery=N4IgZglgNgLgpgJxALhCAvkA&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVNQAooAueAZwxi1QHMBKMgNxy2AG4AoUSWBFdNjwFiZStTqN4LNuyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdFETRoAe2iRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare function fn(a: string): void;
declare function fn(a: string): void;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/unified-signatures": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

These signatures should be reported

### Actual Result

Nothing is reported

### Additional Info

Found this by mistake when reviewing #12656 

Not sure this example represents real code though, so we can skip fixing this if it's not worth the effort

**Comment by bradzacher** [comment]
It's technically valid code but also not really code someone would write.

I think its not supported just because it's a different ast node (`TSDeclareFunction`) to what we look for?

**Comment by StyleShit** [comment]
It reproduces also without declare:

```ts
function fn(a: string): void;
function fn(a: string): void {}
```

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&esQuery=N4IgZglgNgLgpgJxALhCAvkA&code=GYVwdgxgLglg9mABMMAKAhgLkQZygJxjAHMBKbANzhgBMBuAKFElgWTS1wKLMupsQBvRAF8gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdFETRoAe2iRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

It's probably not real code, yeah. But if you have a huge block of overloads with jsdocs maybe you can miss such case? Idk

**Comment by bradzacher** [comment]
That particular case is actually kinda working as expected - because there is only one signature declared!

The implementation does not count as a signature.

But the rule could probably warn about that case cos the signature is not necessary.
